src/xb-silo-query.c

```diff
@@ -289,6 +299,53 @@ xb_silo_query_with_root (XbSilo *self, XbNode *n, const gchar *xpath, guint limi
 	return g_steal_pointer (&results);
 }
 
+/**
+ * xb_silo_query_with_root: (skip)
+ * @self: a #XbSilo
+ * @n: (allow-none): a #XbNode
+ * @xpath: an XPath, e.g. `/components/component[@type=desktop]/id[abe.desktop]`
+ * @limit: maximum number of results to return, or 0 for "all"
+ * @error: the #GError, or %NULL
+ *
+ * Searches the silo using an XPath query, returning up to @limit results.
+ *
+ * It is safe to call this function from a different thread to the one that
+ * created the #XbSilo.
+ *
+ * Please note: Only a subset of XPath is supported.
+ *
+ * Returns: (transfer container) (element-type XbNode): results, or %NULL if unfound
+ *
+ * Since: 0.1.0
+ **/
+GPtrArray *
+xb_silo_query_with_root (XbSilo *self, XbNode *n, const gchar *xpath, guint limit, GError **error)
+{
+	return silo_query_with_root (self, n, xpath, limit, FALSE, error);
+}
+
+/**
+ * xb_silo_query_sn_with_root: (skip)
+ * @self: a #XbSilo
+ * @n: (allow-none): a #XbNode
+ * @xpath: an XPath, e.g. `/components/component[@type=desktop]/id[abe.desktop]`
+ * @limit: maximum number of results to return, or 0 for "all"
+ * @error: the #GError, or %NULL
+ *
+ * A version of xb_silo_query_with_root() which returns results as #XbSiloNodes,
+ * rather than as #XbNodes. This is intended to be used internally to save on
+ * intermediate #XbNode allocations.
```

Comment:
if it's internal, we shouldn't export it

Reply:
How do you mean? It’s declared in `xb-silo-query-private.h` and isn’t listed in `libxmlb.map`.

Reply:
my mistake, sorry.